### ISSUE TITLE
Add Component Matchers for Various Drop Configurations

### ISSUE THREAD

**Issue by legendblade** [legendblade/CraftingHarmonics#49]
For #34, #41, #42, and #61 , add in the following matchers

|Matcher|Mob Drop|Block Drop|Mob Shed|Potion Effect|Target Release|
|---------------------------------------------|--------------|---------------|---------------|-----------------|---------------------|---------|
|minLootingLevel (int)|Yes||||1.3.9|
|maxLootingLevel (int)|Yes||||1.3.9|
|minedWith/killedWith (Item minedWith/killedWith, boolean consume, float damagePer, NBT nbt, boolean fuzzyNbt)|Yes|Yes|||1.3.12|
|itemInOffhand/attackerHasOffhand (Item itemInOffhand/attackerHasOffhand, boolean consumeOffhand, float damageOffhandPer, NBT nbtOffhand, boolean fuzzyNbtOffand)|Yes|Yes|||1.3.12|
|itemInInventory/playerHasInventory (Item itemInInventory/playerHasInventory , boolean consumeInventory, float damageInventoryPer, NBT nbtInventory, boolean fuzzyNbtInventory)|Yes|Yes||Yes|1.3.12|
|inDimension (list: dimension IDs)|Yes|Yes|Yes|Yes|1.3.12|
|inBiome (list: biomes)|Yes|Yes|Yes|Yes|1.3.12|
|inBiomeType (list: types)|Yes|Yes|Yes|Yes|1.3.12|
|minMobHealth (float)|||Yes|Yes|1.3.12|
|maxMobHealth (float)|||Yes|Yes|1.3.12|
|damageOnShed (float damageOnShed, String sheddingDamageType, boolean sheddingDamageIsUnblockable, boolean sheddingDamageIsAbsolute, boolean sheddingDamageIsCreative)|||Yes||1.3.12|
|minLightLevel (int)|Yes|Yes|Yes|Yes|1.3.12|
|maxLightLevel (int)|Yes|Yes|Yes|Yes|1.3.12|
|targetHasPotionEffect: (potion ID)|Yes||Yes|Yes|1.3.12|
|playerHasPotionEffect: (potion ID) / attackerHasPotionEffect (for mob drops)|Yes|Yes|||1.3.12|
|silkTouch (boolean)||Yes|||1.3.12|
|minedByPlayer (boolean)||Yes|||1.3.12|
|minedByExplosion (boolean)||Yes|||1.3.12|
|minFortuneLevel (int)||Yes|||1.3.12|
|maxFortuneLevel (int)||Yes|||1.3.12|
|minHeight (int)|Yes|Yes|Yes|Yes|1.3.12|
|maxHeight (int)|Yes|Yes|Yes|Yes|1.3.12|
|chance (chance %)|Yes|Yes|Yes|Yes|1.3.12|
|not (any other matchers)|Yes|Yes|Yes|Yes|1.3.12|
|or (any other matchers)|Yes|Yes|Yes|Yes|1.3.12|
|and (any other matchers)|Yes|Yes|Yes|Yes|1.3.12|
|aboveBlock (block with nbt)|Yes|Yes|Yes|Yes|1.3.13|
|aroundMobs (List what, int min, int max, int dist)|Yes|Yes|Yes|Yes|1.3.13|
|aroundBlocks (List what, int min, int max, int dist)|Yes|Yes|Yes|Yes|1.3.13|
|minTimeOfDay (int)|Yes|Yes|Yes|Yes|1.3.13|
|maxTimeOfDay (int)|Yes|Yes|Yes|Yes|1.3.13|
|moonPhase (phase)|Yes|Yes|Yes|Yes|1.3.13|
|insideBlock: (block with nbt)|Yes||Yes|Yes|1.3.13|
|mobHasMainhand (Item mobHasMainhand, NBT mobMainhandNbt, boolean mobMainhandFuzzyNbt)|Yes||Yes|Yes|1.3.13|
|mobHasOffhand (Item mobHasOffhand, NBT mobOffhandNbt, boolean mobOffhandFuzzyNbt)|Yes||Yes|Yes|1.3.13|
|damagePlayer (float damagePlayer, String playerDamageType, boolean playerDamageIsUnblockable, boolean playerDamageIsAbsolute, boolean playerDamageIsCreative)|Yes|Yes|||1.3.13|
|minScore (String scoreboard ID, int min)|Yes|Yes||Yes|1.3.13|
|maxScore (String scoreboard ID, int max)|Yes|Yes||Yes|1.3.13|
|targetRiding|||Yes|Yes|1.4.2|
|isRaining|Yes|Yes|Yes|Yes|1.4.2|
|isStorming|Yes|Yes|Yes|Yes|1.4.2|
|playerHasAchievement (list: achievement ID)|Yes|Yes||Yes|1.4.2|
|playerHasStatistic (list: statistic ID, min, max)|Yes|Yes||Yes|1.4.2|
|Seasons.inSeason (list: season)|Yes|Yes|Yes|Yes|???|
|Questing.questActive (list: quest ID, status)|Yes|Yes||Yes|???|
|Questing.hasPoints (list: pool ID, min, max)|Yes|Yes||Yes|???|

**Comment by Thiana** [comment]
A few things:

(1) toolFortuneLevel should be for mobs as well (Looting).
(2) aroundMobs could use a notAroundMobs complement. 
(3) item* should allow detection of nbt/partial-nbt.
(4) As a more generic option than leaderboard checking, allow matching based on http://minecraft.gamepedia.com/Statistics since that is where the leaderboards pull their data from.
(5) Perhaps add achievement checking as well?

How will 'conditions' like consume/damagePer work with the matchers?

Thanks.

**Comment by legendblade** [comment]
So... in a semi-joking manner...

1. I didn't include looting in there since it was already done and working with mob drops, which, was the only one it'd apply to.
2. Yes, it could.  Or, you know... {mob: "Cow", min: 0: max: 0, distance: 10} :P
3. If I must.  :)
4. You really like giving me work...
5. See 3, and 4.

(updating the list now)

Regarding consume and damagePer, they'll work the same as they currently do.  Basically, a lot of the mob drop code is already "good to go" with this, I just need to take it out of being mob-drop specific and make it able to be used for mob sheds and block drops without duplicating code, if that makes sense.

**Comment by Thiana** [comment]
#1 Ok.

#2 There is no #2, you just imagined it ;)

#3 Well not must, but it would be rather useful :)

#4 I may or may not use. The problem is a lot of the uses I can see would depend on the change in the value rather than the fixed value. ie, when the set is applied it would snapshot the statistic and then the check would compare (origVal - currVal OP checkVal).

I can see the fixed value being useful in other types of modpacks though.

I actually wouldnt use #5  myself. The minecraft achievement tracker seems to be far too fond of resetting itself. I was just mentioning it for completeness ;-P

**Comment by legendblade** [comment]
Hey, just because you're 40 hours into the map, maybe you haven't opened your inventory yet... <.<

So, for 4... I'd probably end up having to do a generic script condition for that one (which I really should do at some point, but... getting useful data into the scripting system is a bit of a chore.  Plus then I'd have to find out what data people wanted to check the delta of between current and when the set was applied...

It'd be a process, but, it could be done at some point.  Probably not during the first release of it though.  :)

**Comment by Thiana** [comment]
Apparently I shouldn't depend on quick tests between meetings. You can simulate reputation type tracking with the minecraft scoreboard since MC 1.8. The 'reputation' just has to be defined first as a new objective (using 'scoreboard objectives add ...'). BQ already has support for it in the standard pack. 

So in short, yes I need scoreboard tracking ;) Thanks.

**Comment by legendblade** [comment]
@Thiana Not sure if you caught it, but I bumped scoreboard tracking up to a 1.3.13 release.  Which is pending on CF right now. :)

**Comment by Thiana** [comment]
I hadn't caught it. I've been working on structures for worldgen. 

Thanks :) And the boolean logic is a very nice bonus :)

Question, what are your plans for 1.9.4? Moving to, or updating in parallel with 1.9? Right now it looks like a little more than half my modlist is on 1.9, two still on 1.8.9 and the rest on 1.9.4.

**Comment by legendblade** [comment]
The boolean logic was mostly just me being lazy... because I realized I was going to have to write a lot of "NotSomethingSomethingSomething" classes otherwise. :)

I'd originally planned on just moving forward to 1.9.4 and backporting any major bugs that were found to the 1.9 version, but, I'm going to have to see how things go.  I've been doing a lot of massive internal rework as part of the 1.4.0 version, because my original structure for everything... wasn't the best in order to support the project going forward.  As such, I'm not sure how easy it'll be to even have 1.9 backports.

But!  If it's not too much work, I'll try and keep both versions updated for at least a little while during the transition?

**Comment by Thiana** [comment]
Looks like several more of my mods updated to 1.9.4 overnight. Since CH+BQ are the 'core' and available/will be available for 1.9.4 I'm going to update as well.

**Comment by legendblade** [comment]
For Seasons support, do you also want a matcher for the calculated temperature at the location, or just the current season itself?  I'm assuming both would be good. :)

**Comment by Thiana** [comment]
I only need the current season at the moment since idea is to enhance the effect of 'Blessings' if the season matches that blessing, ie Spring/Summer for the Life deity, Winter for the Spirit/Death deity. I might also tie mob aggression/power to the seasons, ie mobs are more desperate (thus powerful) in non-plentiful seasons. Still debating that part.

I'm actually not sure that the calculated temperature would be that useful? You already implemented inBiomeType which gives us a temperature range option... The current location would just seem to be a duplication of that, unless you want to be really nasty to the player and tell them to find a spot within a biome that is in a particular temperature range. I think that would be a little extreme for anything outside of Blood and Bones ;)
